docs/developers/gas-and-fees/user-defined-smart-contracts.md

```diff
@@ -0,0 +1,251 @@
+Documentation in this section is preliminary and subject to change.
+:::
+
+Before moving forward, make sure you first have a look over the following:
+ - [Asynchronous calls between contracts](/technology/the-elrond-wasm-vm/#asynchronous-calls-between-contracts)
+ - [Asynchronous calls (Rust framework)](/developers/developer-reference/elrond-wasm-contract-calls/#asynchornous-calls)
+ - [Callbacks (Rust framework)](/developers/developer-reference/elrond-wasm-annotations/#callbacks)
+
+Suppose we have two contracts: `A` and `B`, where `A::foo(addressOfB)` calls `B::bar()`.
+
+Let's deploy the contracts `A` and `B`:
+
+```
+$ erdpy --verbose contract deploy --bytecode=./a.wasm \
+ --recall-nonce --gas-limit=5000000 \
+ --pem=./testnet/wallets/users/alice.pem \
+ --send --wait-result --outfile=a.json
+
+$ erdpy --verbose contract deploy --bytecode=./b.wasm \
+ --recall-nonce --gas-limit=5000000 \
+ --pem=./testnet/wallets/users/alice.pem \
+ --send --wait-result --outfile=b.json
+```
+
+Assuming `A` is deployed at `erd1qqqqqqqqqqqqqpgqfzydqmdw7m2vazsp6u5p95yxz76t2p9rd8ss0zp9ts`, and `B` is deployed at `erd1qqqqqqqqqqqqqpgqj5zftf3ef3gqm3gklcetpmxwg43rh8z2d8ss2e49aq`, let's **simulate** `A::foo(addressOfB)` (at first, pass a _large-enough_ or maximum `gas-limit`):
+
+```
+$ export hexAddressOfB=0x$(erdpy wallet bech32 --decode erd1qqqqqqqqqqqqqpgqj5zftf3ef3gqm3gklcetpmxwg43rh8z2d8ss2e49aq)
+
+$ erdpy --verbose contract call erd1qqqqqqqqqqqqqpgqfzydqmdw7m2vazsp6u5p95yxz76t2p9rd8ss0zp9ts \
+ --pem=./testnet/wallets/users/alice.pem \
+ --function=foo\
+ --recall-nonce --gas-limit=50000000\
+ --arguments ${hexAddressOfB}\
+ --simulate
+```
+
+In the output, look for the simulated cost (as above):
+
+```
+"txSimulation": {
+    ...
+    "cost": {
+        "txGasUnits": 3473900,
+        ...
+    }
+}
+```
+
+The simulated cost represents the **actual gas cost** for invoking `A::foo()`, `B::bar()` and `A::callBack()`.
+
+**However, the simulated cost above isn't the value we are going to use as `gasLimit`.** If we were to do so, we would be presented the error `not enough gas`.
+
+The Elrond VM, while hitting the call to `B::bar()` - also called an _asyncCall breakpoint_ - before actually invoking the function, inspects the **_on-the-spot_ remaining gas** and **temporarily locks (reserves) a portion of it**, to allow for the execution of `A::callBack()` in the end.
+
+With respect to the [VM Gas Schedule](https://github.com/ElrondNetwork/elrond-config-mainnet/tree/master/gasSchedules), the aforementioned **_on-the-spot_ remaining gas** has to satisfy the following (in order for the **temporary gas lock reservation** to succeed):
+
+```
+onTheSpotRemainingGas > gasToLockForCallback
+
+gasToLockForCallback = 
+    costOf(AsyncCallStep) + 
+    costOf(AsyncCallbackGasLock) + 
+    codeSizeOf(callingContract) * costOf(AoTPreparePerByte)
+```
+
+:::note
+Multiple asynchronous calls employ multiple temporary gas locks, one for each callback function.
```

Comment:
Multiple asynchronous calls are not yet supported by the VM, not sure if we should mention this here.